File: backend/domain_monitor/source_adapters.py
from __future__ import annotations

import csv
import gzip
import io
import json
import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable, Iterator, Optional, Protocol

import requests

import source_config
from source_config import SourceSettings

logger = logging.getLogger(__name__)

USER_AGENT = "SEO-Tool-Platform-DomainRadar/1.0"

STATUS_ACTIVE = "Active"
STATUS_CONFIGURED = "Configured"
STATUS_NOT_CONFIGURED = "Not Configured"
STATUS_DISABLED = "Disabled"
STATUS_FAILED = "Failed"


@dataclass
class SourceReport:
    """What one adapter did on this run. Surfaced verbatim in the UI."""

    kind: str
    name: str
    label: str
    status: str
    configured: bool
    enabled: bool
    raw_count: int = 0
    detail: str = ""
    error: Optional[str] = None


class DomainSource(Protocol):
    """A candidate-domain provider.

    `fetch_domains` yields raw strings; normalisation, validation and
    deduplication all happen downstream in the collector, so an adapter never
    has to care about domain syntax.
    """

    kind: str
    name: str
    label: str

    def is_configured(self) -> bool: ...

    def describe(self) -> str: ...

    def fetch_domains(self) -> Iterable[str]: ...


class _BaseSource:
    kind = "base"
    name = "base"
    label = "Base"

    def __init__(self, settings: SourceSettings) -> None:
        self.settings = settings

    def is_configured(self) -> bool:
        return True

    def describe(self) -> str:
        return ""

    def fetch_domains(self) -> Iterable[str]:
        return []


def _cells(line: str, column: int) -> Iterator[str]:
    """Yield the candidate cell from one TXT or CSV line, skipping comments."""
    text = line.strip()
    if not text or text.startswith("#") or text.startswith(";"):
        return
    if "," not in text and "\t" not in text and ";" not in text:
        yield text
        return
    parts = [p.strip().strip('"') for p in text.replace("\t", ",").replace(";", ",").split(",")]
    if column < len(parts) and parts[column]:
        yield parts[column]
    elif parts and parts[0]:
        yield parts[0]


def _cells_first(line: str) -> Iterator[str]:
    """First column of a TXT or CSV line."""
    return _cells(line, 0)


class ManualFileSource(_BaseSource):
    """Domains an admin uploaded or pasted from the dashboard.

    Reads a TXT/CSV file inside the configured manual directory. This is a
    first-class source, but it is empty until someone imports something.
    """

    kind = source_config.KIND_MANUAL
    name = "manual"
    label = "Manual Import"

    @property
    def path(self) -> Path:
        root = self.settings.manual_dir.resolve()
        # Basename only, so a configured filename can never escape the folder.
        return (root / Path(self.settings.manual_file).name).resolve()

    def is_configured(self) -> bool:
        return self.path.exists() and self.path.stat().st_size > 0

    def describe(self) -> str:
        if not self.path.exists():
            return "No imported list yet — upload a TXT or CSV from the dashboard"
        return f"{self.path.name} ({self.path.stat().st_size:,} bytes)"

    def fetch_domains(self) -> Iterator[str]:
        path = self.path
        if not path.exists():
            return
        with path.open("r", encoding="utf-8", errors="ignore") as handle:
            for line in handle:
                yield from _cells(line, 0)


class ZoneFileSource(_BaseSource):
    """Registry zone files as a candidate source.

    Zone files list registered domains; they carry no expiry information, so
    every candidate still goes through RDAP for its lifecycle. Supports .txt,
    .zone, .csv and .gz.
    """

    kind = source_config.KIND_ZONE
    name = "zone"
    label = "Zone File"

    SUFFIXES = (".txt", ".zone", ".csv", ".gz")

    def is_configured(self) -> bool:
        directory = self.settings.zone_directory
        return bool(directory and directory.is_dir() and self._files())

    def _files(self) -> list[Path]:
        directory = self.settings.zone_directory
        if not directory or not directory.is_dir():
            return []
        found = [
            path
            for path in sorted(directory.iterdir())
            if path.is_file() and path.suffix.lower() in self.SUFFIXES
        ]
        return found[: self.settings.zone_max_files]

    def describe(self) -> str:
        directory = self.settings.zone_directory
        if not directory:
            return "Zone source not configured (set ZONE_FILE_DIRECTORY)"
        if not directory.is_dir():
            return f"Zone directory not found: {directory}"
        files = self._files()
        if not files:
            return f"No zone files in {directory}"
        total = sum(path.stat().st_size for path in files)
        return f"{len(files)} zone file(s) in {directory.name}, {total / 1_048_576:.1f} MB"

    @staticmethod
    def _inner_suffix(path: Path) -> str:
        """Effective format, looking through a .gz wrapper."""
        if path.suffix.lower() == ".gz":
            return Path(path.stem).suffix.lower() or ".txt"
        return path.suffix.lower()

    def fetch_domains(self) -> Iterator[str]:
        """Stream candidates line by line; a zone file never lands in memory whole."""
        for path in self._files():
            # Zone syntax and delimited lists need different parsers, so pick by
            # the effective extension rather than guessing per line.
            parse = self._zone_line if self._inner_suffix(path) == ".zone" else _cells_first
            try:
                if path.suffix.lower() == ".gz":
                    handle: Any = gzip.open(path, "rt", encoding="utf-8", errors="ignore")
                else:
                    handle = path.open("r", encoding="utf-8", errors="ignore")
                with handle:
                    for line in handle:
                        yield from parse(line)
            except OSError as exc:
                # One unreadable file must not abort the remaining zone files.
                logger.warning("[source:zone] could not read %s: %s", path.name, exc)

    @staticmethod
    def _zone_line(line: str) -> Iterator[str]:
        """Extract the owner name from a zone record, or the whole line.

        Handles both a bare domain list and real zone syntax such as
        `example.com. 172800 IN NS ns1.example.com.`
        """
        text = line.strip()
        if not text or text.startswith((";", "#", "$")):
            return
        first = text.split()[0]
        if first:
            yield first.rstrip(".")


class ExternalFeedSource(_BaseSource):
    """A configured HTTP domain feed.

    Only for feeds whose operator permits automated retrieval. There is no HTML
    parsing, no login flow and no anti-bot circumvention here: the adapter does
    one authenticated GET and reads text, CSV or JSON.
    """

    kind = source_config.KIND_FEED
    name = "feed"
    label = "External Feed"

    def is_configured(self) -> bool:
        return bool(self.settings.feed_url)

    def describe(self) -> str:
        if not self.settings.feed_url:
            return "Feed not configured (set DOMAIN_FEED_URL)"
        host = self.settings.feed_url.split("/")[2] if "//" in self.settings.feed_url else self.settings.feed_url
        auth = "with API key" if self.settings.feed_api_key else "no credential"
        return f"{host} ({self.settings.feed_format}, {auth})"

    def fetch_domains(self) -> Iterator[str]:
        url = self.settings.feed_url
        if not url:
            return

        headers = {"User-Agent": USER_AGENT, "Accept": "text/plain, text/csv, application/json"}
        if self.settings.feed_api_key:
            headers["Authorization"] = f"Bearer {self.settings.feed_api_key}"

        try:
            response = requests.get(
                url, headers=headers, timeout=self.settings.timeout, stream=True
            )
            response.raise_for_status()
            body = response.raw.read(self.settings.max_fetch_bytes, decode_content=True) or b""
        except requests.RequestException as exc:
            # Raised so the collector can mark this source Failed and continue.
            raise RuntimeError(f"Feed request failed: {exc}") from exc

        text = body.decode("utf-8", errors="ignore")
        fmt = self.settings.feed_format
        if fmt == "auto":
            stripped = text.lstrip()
            fmt = "json" if stripped[:1] in ("{", "[") else "csv" if "," in stripped[:400] else "text"

        if fmt == "json":
            yield from self._from_json(text)
        else:
            for line in text.splitlines():
                yield from _cells(line, self.settings.feed_column)

    def _from_json(self, text: str) -> Iterator[str]:
        """Pull domains out of a JSON body.

        DOMAIN_FEED_JSON_PATH names the list (dotted) and, after a colon, the
        field holding the domain — e.g. `data.items:domain`.
        """
        try:
            payload = json.loads(text)
        except ValueError as exc:
            raise RuntimeError(f"Feed returned invalid JSON: {exc}") from exc

        spec = self.settings.feed_json_path
        list_path, _, field_name = spec.partition(":")

        node: Any = payload
        for part in [p for p in list_path.split(".") if p]:
            if isinstance(node, dict):
                node = node.get(part)
            else:
                node = None
                break

        if node is None and isinstance(payload, dict):
            node = next((v for v in payload.values() if isinstance(v, list)), None)
        if node is None:
            node = payload

        if not isinstance(node, list):
            raise RuntimeError("Feed JSON did not contain a list of domains")

        for entry in node:
            if isinstance(entry, str):
                yield entry
            elif isinstance(entry, dict):
                if field_name and field_name in entry:
                    yield str(entry[field_name])
                else:
                    for key in ("domain", "name", "host", "hostname", "fqdn"):
                        if key in entry:
                            yield str(entry[key])
                            break


class DatabaseWatchlistSource(_BaseSource):
    """Re-queues domains already flagged by an admin.

    Keeps watchlisted candidates refreshed even if the source that originally
    discovered them has gone away.
    """

    kind = source_config.KIND_WATCHLIST
    name = "watchlist"
    label = "Watchlist"

    def is_configured(self) -> bool:
        return True

    def describe(self) -> str:
        return "Domains flagged in the dashboard"

    def fetch_domains(self) -> Iterator[str]:
        import storage

        storage.migrate()
        for row in storage.watchlisted_domains():
            yield row["domain"]


class DemoFixtureSource(_BaseSource):
    """Sample domains for local development only.

    Loads nothing unless DOMAIN_USE_DEMO_DATA is explicitly true, so production
    can never be seeded with fixtures by accident.
    """

    kind = source_config.KIND_DEMO
    name = "demo"
    label = "Demo Fixture"

    def is_configured(self) -> bool:
        return self.settings.demo_enabled and self.settings.demo_file.exists()

    def describe(self) -> str:
        if not self.settings.demo_enabled:
            return "Disabled (set DOMAIN_USE_DEMO_DATA=true to load fixtures)"
        if not self.settings.demo_file.exists():
            return f"Fixture file missing: {self.settings.demo_file}"
        return f"Development fixtures from {self.settings.demo_file.name}"

    def fetch_domains(self) -> Iterator[str]:
        if not self.settings.demo_enabled:
            return
        path = self.settings.demo_file
        if not path.exists():
            return
        logger.warning(
            "[source:demo] loading development fixtures from %s — not for production",
            path,
        )
        with path.open("r", encoding="utf-8", errors="ignore") as handle:
            for line in handle:
                yield from _cells(line, 0)


class InlineListSource(_BaseSource):
    """Domains supplied directly by a request, e.g. a single-domain recheck."""

    kind = "inline"
    name = "request"
    label = "Request"

    def __init__(self, settings: SourceSettings, domains: Iterable[str]) -> None:
        super().__init__(settings)
        self._domains = list(domains)

    def is_configured(self) -> bool:
        return bool(self._domains)

    def describe(self) -> str:
        return f"{len(self._domains)} domain(s) supplied by the request"

    def fetch_domains(self) -> Iterable[str]:
        return list(self._domains)


class Crawl4AIDomainSource(_BaseSource):
    """Configured Crawl4AI source.

    The discovery interface stays synchronous; Crawl4AI does its async work
    internally and returns cached candidates here.
    """

    kind = source_config.KIND_CRAWL4AI
    label = "Crawl4AI"

    def __init__(self, settings: SourceSettings, source_id: str) -> None:
        super().__init__(settings)
        import crawl4ai_source

        self._source_id = source_id
        self._config = next(
            (row for row in crawl4ai_source.load_source_configs() if row.id == source_id),
            None,
        )
        self.name = self._config.name if self._config else source_id

    def is_configured(self) -> bool:
        return bool(self._config and self._config.url)

    def describe(self) -> str:
        if not self._config:
            return "No crawler source configured"
        return f"{self._config.url} · {self._config.max_pages} page cap"

    def fetch_domains(self) -> Iterator[str]:
        if not self._config or not self._config.enabled:
            return
        import crawl4ai_source

        result = crawl4ai_source.crawl_source(self._config)
        for domain in result.domains:
            yield domain


ADAPTERS: dict[str, type[_BaseSource]] = {
    source_config.KIND_MANUAL: ManualFileSource,
    source_config.KIND_ZONE: ZoneFileSource,
    source_config.KIND_FEED: ExternalFeedSource,
    source_config.KIND_WATCHLIST: DatabaseWatchlistSource,
    source_config.KIND_DEMO: DemoFixtureSource,
}


def build_adapter(kind: str, settings: SourceSettings) -> Optional[_BaseSource]:
    builder = ADAPTERS.get(kind)
    return builder(settings) if builder else None


def build_adapters(kind: str, settings: SourceSettings) -> list[_BaseSource]:
    if kind == source_config.KIND_CRAWL4AI:
        import crawl4ai_source

        return [
            Crawl4AIDomainSource(settings, source.id)
            for source in crawl4ai_source.load_source_configs()
            if source.enabled
        ]
    adapter = build_adapter(kind, settings)
    return [adapter] if adapter else []


def all_adapters(settings: SourceSettings) -> list[_BaseSource]:
    """Every known adapter, configured or not, for the status panel."""
    return [builder(settings) for builder in ADAPTERS.values()]

File: node-backend/domainMonitor/sourceAdapters.js
"use strict";

// Faithful Node port of backend/domain_monitor/source_adapters.py
const fs = require("fs");
const path = require("path");
const zlib = require("zlib");
const axios = require("axios");

const sourceConfig = require("./sourceConfig");

const USER_AGENT = "SEO-Tool-Platform-DomainRadar/1.0";

const STATUS_ACTIVE = "Active";
const STATUS_CONFIGURED = "Configured";
const STATUS_NOT_CONFIGURED = "Not Configured";
const STATUS_DISABLED = "Disabled";
const STATUS_FAILED = "Failed";

/** What one adapter did on this run. Surfaced verbatim in the UI. */
class SourceReport {
  constructor({
    kind,
    name,
    label,
    status,
    configured,
    enabled,
    raw_count = 0,
    detail = "",
    error = null,
  }) {
    this.kind = kind;
    this.name = name;
    this.label = label;
    this.status = status;
    this.configured = configured;
    this.enabled = enabled;
    this.raw_count = raw_count;
    this.detail = detail;
    this.error = error;
  }
}

/**
 * A candidate-domain provider.
 *
 * `fetchDomains` yields raw strings; normalisation, validation and
 * deduplication all happen downstream in the collector, so an adapter never
 * has to care about domain syntax.
 */
class DomainSource {
  constructor(settings) {
    this.settings = settings;
  }

  isConfigured() {
    return true;
  }

  describe() {
    return "";
  }

  fetchDomains() {
    return [];
  }
}
DomainSource.prototype.kind = "base";
DomainSource.prototype.name = "base";
DomainSource.prototype.label = "Base";

/**
 * Yield the candidate cell from one TXT or CSV line, skipping comments.
 * @returns {IterableIterator<string>}
 */
function* cells(line, column) {
  const text = line.trim();
  if (!text || text.startsWith("#") || text.startsWith(";")) {
    return;
  }
  if (!text.includes(",") && !text.includes("\t") && !text.includes(";")) {
    yield text;
    return;
  }
  const parts = text
    .replace(/\t/g, ",")
    .replace(/;/g, ",")
    .split(",")
    .map((p) => p.trim().replace(/^"+|"+$/g, ""));
  if (column < parts.length && parts[column]) {
    yield parts[column];
  } else if (parts.length && parts[0]) {
    yield parts[0];
  }
}

/** First column of a TXT or CSV line. */
function* cellsFirst(line) {
  yield* cells(line, 0);
}

/** Read a text file line by line as an iterator, matching Python's line iteration. */
function* readLines(text) {
  // Python iterates a file yielding lines including trailing newline; cells()
  // strips whitespace anyway, so splitting on newlines is equivalent.
  for (const line of text.split(/\r?\n/)) {
    yield line;
  }
}

class ManualFileSource extends DomainSource {
  get path() {
    const root = path.resolve(this.settings.manual_dir);
    // Basename only, so a configured filename can never escape the folder.
    return path.resolve(root, path.basename(this.settings.manual_file));
  }

  isConfigured() {
    try {
      const st = fs.statSync(this.path);
      return st.isFile() && st.size > 0;
    } catch {
      return false;
    }
  }

  describe() {
    if (!fs.existsSync(this.path)) {
      return "No imported list yet — upload a TXT or CSV from the dashboard";
    }
    const size = fs.statSync(this.path).size;
    return `${path.basename(this.path)} (${size.toLocaleString("en-US")} bytes)`;
  }

  *fetchDomains() {
    const p = this.path;
    if (!fs.existsSync(p)) {
      return;
    }
    const text = fs.readFileSync(p, "utf-8");
    for (const line of readLines(text)) {
      yield* cells(line, 0);
    }
  }
}
ManualFileSource.prototype.kind = sourceConfig.KIND_MANUAL;
ManualFileSource.prototype.name = "manual";
ManualFileSource.prototype.label = "Manual Import";

const ZONE_SUFFIXES = [".txt", ".zone", ".csv", ".gz"];

/** Extract the owner name from a zone record, or the whole line. */
function* zoneLine(line) {
  const text = line.trim();
  if (!text || text.startsWith(";") || text.startsWith("#") || text.startsWith("$")) {
    return;
  }
  const first = text.split(/\s+/)[0];
  if (first) {
    yield first.replace(/\.+$/, "");
  }
}

class ZoneFileSource extends DomainSource {
  isConfigured() {
    const directory = this.settings.zone_directory;
    return Boolean(directory && isDir(directory) && this._files().length);
  }

  _files() {
    const directory = this.settings.zone_directory;
    if (!directory || !isDir(directory)) {
      return [];
    }
    const found = fs
      .readdirSync(directory)
      .sort()
      .map((entry) => path.join(directory, entry))
      .filter((p) => {
        try {
          return (
            fs.statSync(p).isFile() &&
            ZONE_SUFFIXES.includes(path.extname(p).toLowerCase())
          );
        } catch {
          return false;
        }
      });
    return found.slice(0, this.settings.zone_max_files);
  }

  describe() {
    const directory = this.settings.zone_directory;
    if (!directory) {
      return "Zone source not configured (set ZONE_FILE_DIRECTORY)";
    }
    if (!isDir(directory)) {
      return `Zone directory not found: ${directory}`;
    }
    const files = this._files();
    if (!files.length) {
      return `No zone files in ${directory}`;
    }
    const total = files.reduce((sum, p) => sum + fs.statSync(p).size, 0);
    return `${files.length} zone file(s) in ${path.basename(directory)}, ${(
      total / 1048576
    ).toFixed(1)} MB`;
  }

  /** Effective format, looking through a .gz wrapper. */
  static _innerSuffix(p) {
    if (path.extname(p).toLowerCase() === ".gz") {
      return path.extname(path.basename(p, path.extname(p))).toLowerCase() || ".txt";
    }
    return path.extname(p).toLowerCase();
  }

  *fetchDomains() {
    for (const p of this._files()) {
      const parse =
        ZoneFileSource._innerSuffix(p) === ".zone" ? zoneLine : cellsFirst;
      let text;
      try {
        if (path.extname(p).toLowerCase() === ".gz") {
          text = zlib.gunzipSync(fs.readFileSync(p)).toString("utf-8");
        } else {
          text = fs.readFileSync(p, "utf-8");
        }
      } catch (exc) {
        // One unreadable file must not abort the remaining zone files.
        console.warn(`[source:zone] could not read ${path.basename(p)}: ${exc}`);
        continue;
      }
      for (const line of readLines(text)) {
        yield* parse(line);
      }
    }
  }
}
ZoneFileSource.prototype.kind = sourceConfig.KIND_ZONE;
ZoneFileSource.prototype.name = "zone";
ZoneFileSource.prototype.label = "Zone File";

class ExternalFeedSource extends DomainSource {
  isConfigured() {
    return Boolean(this.settings.feed_url);
  }

  describe() {
    if (!this.settings.feed_url) {
      return "Feed not configured (set DOMAIN_FEED_URL)";
    }
    const host = this.settings.feed_url.includes("//")
      ? this.settings.feed_url.split("/")[2]
      : this.settings.feed_url;
    const auth = this.settings.feed_api_key ? "with API key" : "no credential";
    return `${host} (${this.settings.feed_format}, ${auth})`;
  }

  async *fetchDomains() {
    const url = this.settings.feed_url;
    if (!url) {
      return;
    }

    const headers = {
      "User-Agent": USER_AGENT,
      Accept: "text/plain, text/csv, application/json",
    };
    if (this.settings.feed_api_key) {
      headers.Authorization = `Bearer ${this.settings.feed_api_key}`;
    }

    let text;
    try {
      const response = await axios.get(url, {
        headers,
        timeout: this.settings.timeout * 1000,
        responseType: "arraybuffer",
        maxContentLength: this.settings.max_fetch_bytes,
        maxBodyLength: this.settings.max_fetch_bytes,
        // Match requests.raise_for_status(): non-2xx must raise.
      });
      let body = Buffer.from(response.data || "");
      if (body.length > this.settings.max_fetch_bytes) {
        body = body.subarray(0, this.settings.max_fetch_bytes);
      }
      text = body.toString("utf-8");
    } catch (exc) {
      // Raised so the collector can mark this source Failed and continue.
      throw new Error(`Feed request failed: ${exc.message || exc}`);
    }

    let fmt = this.settings.feed_format;
    if (fmt === "auto") {
      const stripped = text.replace(/^\s+/, "");
      const head = stripped.slice(0, 1);
      fmt =
        head === "{" || head === "["
          ? "json"
          : stripped.slice(0, 400).includes(",")
          ? "csv"
          : "text";
    }

    if (fmt === "json") {
      yield* this._fromJson(text);
    } else {
      for (const line of text.split(/\r?\n/)) {
        yield* cells(line, this.settings.feed_column);
      }
    }
  }

  /**
   * Pull domains out of a JSON body.
   *
   * DOMAIN_FEED_JSON_PATH names the list (dotted) and, after a colon, the
   * field holding the domain — e.g. `data.items:domain`.
   */
  *_fromJson(text) {
    let payload;
    try {
      payload = JSON.parse(text);
    } catch (exc) {
      throw new Error(`Feed returned invalid JSON: ${exc.message || exc}`);
    }

    const spec = this.settings.feed_json_path || "";
    const colon = spec.indexOf(":");
    const listPath = colon === -1 ? spec : spec.slice(0, colon);
    const fieldName = colon === -1 ? "" : spec.slice(colon + 1);

    let node = payload;
    for (const part of listPath.split(".").filter((p) => p)) {
      if (node && typeof node === "object" && !Array.isArray(node)) {
        node = Object.prototype.hasOwnProperty.call(node, part) ? node[part] : undefined;
      } else {
        node = null;
        break;
      }
    }

    if ((node === null || node === undefined) && isPlainObject(payload)) {
      node = Object.values(payload).find((v) => Array.isArray(v));
      if (node === undefined) node = null;
    }
    if (node === null || node === undefined) {
      node = payload;
    }

    if (!Array.isArray(node)) {
      throw new Error("Feed JSON did not contain a list of domains");
    }

    for (const entry of node) {
      if (typeof entry === "string") {
        yield entry;
      } else if (isPlainObject(entry)) {
        if (fieldName && Object.prototype.hasOwnProperty.call(entry, fieldName)) {
          yield String(entry[fieldName]);
        } else {
          for (const key of ["domain", "name", "host", "hostname", "fqdn"]) {
            if (Object.prototype.hasOwnProperty.call(entry, key)) {
              yield String(entry[key]);
              break;
            }
          }
        }
      }
    }
  }
}
ExternalFeedSource.prototype.kind = sourceConfig.KIND_FEED;
ExternalFeedSource.prototype.name = "feed";
ExternalFeedSource.prototype.label = "External Feed";

class DatabaseWatchlistSource extends DomainSource {
  isConfigured() {
    return true;
  }

  describe() {
    return "Domains flagged in the dashboard";
  }

  *fetchDomains() {
    const storage = require("./storage");
    storage.migrate();
    for (const row of storage.watchlistedDomains()) {
      yield row.domain;
    }
  }
}
DatabaseWatchlistSource.prototype.kind = sourceConfig.KIND_WATCHLIST;
DatabaseWatchlistSource.prototype.name = "watchlist";
DatabaseWatchlistSource.prototype.label = "Watchlist";

class DemoFixtureSource extends DomainSource {
  isConfigured() {
    return this.settings.demo_enabled && fs.existsSync(this.settings.demo_file);
  }

  describe() {
    if (!this.settings.demo_enabled) {
      return "Disabled (set DOMAIN_USE_DEMO_DATA=true to load fixtures)";
    }
    if (!fs.existsSync(this.settings.demo_file)) {
      return `Fixture file missing: ${this.settings.demo_file}`;
    }
    return `Development fixtures from ${path.basename(this.settings.demo_file)}`;
  }

  *fetchDomains() {
    if (!this.settings.demo_enabled) {
      return;
    }
    const p = this.settings.demo_file;
    if (!fs.existsSync(p)) {
      return;
    }
    console.warn(
      `[source:demo] loading development fixtures from ${p} — not for production`
    );
    const text = fs.readFileSync(p, "utf-8");
    for (const line of readLines(text)) {
      yield* cells(line, 0);
    }
  }
}
DemoFixtureSource.prototype.kind = sourceConfig.KIND_DEMO;
DemoFixtureSource.prototype.name = "demo";
DemoFixtureSource.prototype.label = "Demo Fixture";

/** Domains supplied directly by a request, e.g. a single-domain recheck. */
class InlineListSource extends DomainSource {
  constructor(settings, domains) {
    super(settings);
    this._domains = Array.from(domains);
  }

  isConfigured() {
    return Boolean(this._domains.length);
  }

  describe() {
    return `${this._domains.length} domain(s) supplied by the request`;
  }

  fetchDomains() {
    return Array.from(this._domains);
  }
}
InlineListSource.prototype.kind = "inline";
InlineListSource.prototype.name = "request";
InlineListSource.prototype.label = "Request";

class Crawl4AIDomainSource extends DomainSource {
  constructor(settings, sourceId) {
    super(settings);
    const crawl4aiSource = require("./crawl4aiSource");
    this._sourceId = sourceId;
    this._config =
      crawl4aiSource.loadSourceConfigs().find((row) => row.id === sourceId) || null;
    this.name = this._config ? this._config.name : sourceId;
  }

  isConfigured() {
    return Boolean(this._config && this._config.url);
  }

  describe() {
    if (!this._config) {
      return "No crawler source configured";
    }
    return `${this._config.url} · ${this._config.max_pages} page cap`;
  }

  async *fetchDomains() {
    if (!this._config || !this._config.enabled) {
      return;
    }
    const crawl4aiSource = require("./crawl4aiSource");
    const result = await crawl4aiSource.crawlSource(this._config);
    for (const domain of result.domains) {
      yield domain;
    }
  }
}
Crawl4AIDomainSource.prototype.kind = sourceConfig.KIND_CRAWL4AI;
Crawl4AIDomainSource.prototype.label = "Crawl4AI";

const ADAPTERS = {
  [sourceConfig.KIND_MANUAL]: ManualFileSource,
  [sourceConfig.KIND_ZONE]: ZoneFileSource,
  [sourceConfig.KIND_FEED]: ExternalFeedSource,
  [sourceConfig.KIND_WATCHLIST]: DatabaseWatchlistSource,
  [sourceConfig.KIND_DEMO]: DemoFixtureSource,
};

function buildAdapter(kind, settings) {
  const Builder = ADAPTERS[kind];
  return Builder ? new Builder(settings) : null;
}

function buildAdapters(kind, settings) {
  if (kind === sourceConfig.KIND_CRAWL4AI) {
    const crawl4aiSource = require("./crawl4aiSource");
    return crawl4aiSource
      .loadSourceConfigs()
      .filter((source) => source.enabled)
      .map((source) => new Crawl4AIDomainSource(settings, source.id));
  }
  const adapter = buildAdapter(kind, settings);
  return adapter ? [adapter] : [];
}

/** Every known adapter, configured or not, for the status panel. */
function allAdapters(settings) {
  return Object.values(ADAPTERS).map((Builder) => new Builder(settings));
}

// --- Small internal helpers ------------------------------------------------

function isDir(p) {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function isPlainObject(v) {
  return v !== null && typeof v === "object" && !Array.isArray(v);
}

module.exports = {
  USER_AGENT,
  STATUS_ACTIVE,
  STATUS_CONFIGURED,
  STATUS_NOT_CONFIGURED,
  STATUS_DISABLED,
  STATUS_FAILED,
  SourceReport,
  DomainSource,
  ManualFileSource,
  ZoneFileSource,
  ExternalFeedSource,
  DatabaseWatchlistSource,
  DemoFixtureSource,
  InlineListSource,
  Crawl4AIDomainSource,
  ADAPTERS,
  buildAdapter,
  buildAdapters,
  allAdapters,
  cells,
  _cells: cells,
  cellsFirst,
};
